### PR TITLE
Fixes #2167

### LINKED ISSUE #2167: Cannot create sim when merging two modules
https://github.com/Open-Systems-Pharmacology/MoBi/issues/1208

### PR DESCRIPTION
**Pull request by msevestre**
Fixes #2167

# Description

So far, the globla molecule container was only using the LAST container. I propose to merge the molecule properties container (global and local)

see https://github.com/Open-Systems-Pharmacology/MoBi/issues/1208 for bug

## Type of change

Please mark relevant options with an `x` in the brackets.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):
- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.
- [ ] Other (please describe):

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Integration tests
- [ ] Unit tests
- [ ] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):

# Questions (if appropriate):

### CODE REVIEW

File: tests/OSPSuite.HelpersForTests/ModuleHelperForSpecs.cs [hunk 1]
```diff
@@ -60,11 +96,15 @@ private SpatialStructure getSpatialStructureModule1()
...
 
-
-         var organismMoleculeProperties = _objectBaseFactory.Create<IContainer>().WithName(MOLECULE_PROPERTIES)
+         var globalMoleculeProperties = _objectBaseFactory.Create<IContainer>().WithName(MOLECULE_PROPERTIES)
             .WithMode(ContainerMode.Logical)
             .WithContainerType(ContainerType.Molecule);
-         organism.Add(organismMoleculeProperties);
+
+         //add two parameters. One will be replaced when merging the second module
+         globalMoleculeProperties.Add(newConstantParameter("P1", 10));
```
Comment (msevestre) [inline tests/OSPSuite.HelpersForTests/ModuleHelperForSpecs.cs:104]: In this use case, P1  and P2 are set in module 1 and P1 and P3 are set in module 2 with a different value for P1. Test is checking that the last value wins

File: tests/OSPSuite.HelpersForTests/SimulationHelperForSpecs.cs [hunk 2]
```diff
@@ -1,25 +1,24 @@
...
 
 namespace OSPSuite.Helpers
 {
-   public class SimulationHelperForSpecs : ModelHelperForSpecs
+   public class SimulationHelperForSpecs
    {
+      private readonly ModelHelperForSpecs _modelHelper;
       private readonly IModelConstructor _modelConstructor;
 
-      public SimulationHelperForSpecs(IObjectBaseFactory objectBaseFactory, IParameterValuesCreator parameterValuesCreator, IInitialConditionsCreator initialConditionsCreator, IObjectPathFactory objectPathFactory, IDimensionFactory dimensionFactory, IModelConstructor modelConstructor,
```
Comment (msevestre) [inline tests/OSPSuite.HelpersForTests/SimulationHelperForSpecs.cs:12]: favor composition over inheritence lol. WTF

**Review by msevestre — COMMENTED** [review]
So this is the solution where we merge the MoleculeProperties (both global and local)

**Comment by msevestre** [comment]
with these changes I can create a simulation with Tumor in one click

image

File: src/OSPSuite.Core/Domain/Services/ModelConstructor.cs [hunk 3]
```diff
@@ -317,13 +317,19 @@ private void createGlobalMoleculeContainers(ModelConfiguration modelConfiguratio
...
+         //once the global properties have been created, we can remove the global container only used as template
+         var globalMoleculeContainer = model.Root.Container(Constants.MOLECULE_PROPERTIES);
+         if (globalMoleculeContainer != null)
+            model.Root.RemoveChild(globalMoleculeContainer);
       }
 
       private void addLocalParametersToMolecule(ModelConfiguration modelConfiguration)
       {
          var (model, simulationConfiguration) = modelConfiguration;
          // retrieve all molecules container defined int the spatial structure
```
Comment (Yuri05) [inline src/OSPSuite.Core/Domain/Services/ModelConstructor.cs:330]: typo in the comment (_...in**t** the spatial structure..._)

**Review by rwmcintosh — APPROVED** [review]
I have the same question as Juri, but that's all.

File: src/OSPSuite.Core/Domain/Services/SpatialStructureMerger.cs [hunk 4]
```diff
@@ -70,6 +70,21 @@ private void createMergedContainerStructureInRoot(IContainer root, SimulationBui
...
+
+
+         var firstGlobalMoleculeContainer = allGlobalMoleculeContainers.FirstOrDefault();
+         var otherGlobalMoleculeContainer = allGlobalMoleculeContainers.Skip(1).ToList();
+
+         if (firstGlobalMoleculeContainer != null)
+         {
+            otherGlobalMoleculeContainer.Each(x => mergeContainers(firstGlobalMoleculeContainer, x));
+            root.Add(firstGlobalMoleculeContainer);
+         }
```
[the remark this answers is not in the document]
Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/SpatialStructureMerger.cs:87]: Well somehow we are doing this for the whole spatial structure where we merge stuff without cloning. So I can only assume that we have a clone already but I am going to check

[the remark this answers is not in the document]
Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/SpatialStructureMerger.cs:87]: ok you are right. We always update using the mapper before modifying which I am not doing here

File: src/OSPSuite.Core/Domain/Services/SpatialStructureMerger.cs [hunk 5]
```diff
@@ -70,6 +70,22 @@ private void createMergedContainerStructureInRoot(IContainer root, SimulationBui
             //make sure we map the container to a model container so that we do not change the original containers
             .Select(mapToModelContainer)
             .Each(mergeTopContainerIntoModel);
+
+         //create the temporary GLOBAL MOLECULE PROPERTIES THAT WILL BE REMOVED AT THE END but used as based for copying
+         var allGlobalMoleculeContainers = allSpatialStructures
+            .Select(x => x.GlobalMoleculeDependentProperties)
+            .Select(mapToModelContainer)
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/SpatialStructureMerger.cs:77]: @Yuri05 @rwmcintosh Using the mapToModel which cloes everything

**Pull request merged**